### ISSUE TITLE
DELETE 204 No Content causes an error at normalizeResponse

### ISSUE THREAD

**Issue by jerel** [ember-cli/ember-ajax#185]
It appears that an empty response on a successful 204 request will throw an error due to it expecting JSON. I'm using Ember Data with `ember-ajax/mixins/ajax-support`.

Error: Assertion Failed: normalizeResponse must return a valid JSON API document: * One or more of the following keys must be present: "data", "errors", "meta".

![stack](https://cloud.githubusercontent.com/assets/322706/19786205/61df501e-9c62-11e6-864f-7f955c6ed2f1.png)

**Comment by alexlafroscia** [comment]
Does the same happen in Ember Data without the mixin? That error sounds like it's coming from the fact that ED is expecting a `json-api` formatted object but is getting back nothing.

**Comment by jerel** [comment]
Ok, I'm baffled. I had worked with it for a while before opening the issue and it appeared to only be introduced by the mixin (this app has been in production and I just now upgraded from ic-ajax to ember-ajax and added the mixin). Now that I went back and tested again everything just works. Not sure if there was some weirdness caused by the ember-cli proxy or something and a restart solved it?

**Comment by jerel** [comment]
This issue actually is valid, in some sense of the word. I had accidentally changed the calling code yesterday which caused the error to disappear. Steps to reproduce: use `AjaxServiceSupport` in application adapter, call `model.destroyRecord()`, and have the server return `204 No Content`. Removing `AjaxServiceSupport` while changing no other code causes the error to go away.

Snippets below shows how the code path diverges.

Standard Ember Data with JSONAPIAdapter:
```
    ajax: function ajax(url, type, options) {
      var adapter = this;

      var requestData = {
        url: url,
        method: type
      };

      return new Promise(function (resolve, reject) {
        var hash = adapter.ajaxOptions(url, type, options);

        hash.success = function (payload, textStatus, jqXHR) {
          // normalizeResponseHelper is never called on this code path
--->      var response = ajaxSuccess(adapter, jqXHR, payload, requestData);
          _ember['default'].run.join(null, resolve, response);
        };

        hash.error = function (jqXHR, textStatus, errorThrown) {
          var responseData = {
            textStatus: textStatus,
            errorThrown: errorThrown
          };
          var error = ajaxError(adapter, jqXHR, requestData, responseData);
          _ember['default'].run.join(null, reject, error);
        };

        adapter._ajaxRequest(hash);
      }, 'DS: RESTAdapter#ajax ' + type + ' to ' + url);
    },
```

Using the mixin:
```
    _makeRequest: function _makeRequest(hash) {
      var _this2 = this;

      var requestData = {
        type: hash.type,
        url: hash.url
      };

      if (isJSONAPIContentType(hash.headers['Content-Type']) && requestData.type !== 'GET') {
        if (typeof hash.data === 'object') {
          hash.data = JSON.stringify(hash.data);
        }
      }

      return new Promise(function (resolve, reject) {
        hash.success = function (payload, textStatus, jqXHR) {
          var response = _this2.handleResponse(jqXHR.status, (0, _emberAjaxUtilsParseResponseHeaders['default'])(jqXHR.getAllResponseHeaders()), payload, requestData);

          pendingRequestCount = pendingRequestCount - 1;

          if ((0, _emberAjaxErrors.isAjaxError)(response)) {
            run.join(null, reject, { payload: payload, textStatus: textStatus, jqXHR: jqXHR, response: response });
          } else {
            // payload is undefined and textStatus is nocontent
            // normalizeResponseHelper ends up being called after this which causes an error due to no "data" or "error" key
--->        run.join(null, resolve, { payload: payload, textStatus: textStatus, jqXHR: jqXHR, response: response });
          }
        };

        hash.error = function (jqXHR, textStatus, errorThrown) {
          runInDebug(function () {
            var message = 'The server returned an empty string for ' + requestData.type + ' ' + requestData.url + ', which cannot be parsed into a valid JSON. Return either null or {}.';
            var validJSONString = !(textStatus === 'parsererror' && jqXHR.responseText === '');
            warn(message, validJSONString, {
              id: 'ds.adapter.returned-empty-string-as-JSON'
            });
          });

          var payload = _this2.parseErrorResponse(jqXHR.responseText) || errorThrown;
          var response = undefined;

          if (errorThrown instanceof Error) {
            response = errorThrown;
          } else if (textStatus === 'timeout') {
            response = new _emberAjaxErrors.TimeoutError();
          } else if (textStatus === 'abort') {
            response = new _emberAjaxErrors.AbortError();
          } else {
            response = _this2.handleResponse(jqXHR.status, (0, _emberAjaxUtilsParseResponseHeaders['default'])(jqXHR.getAllResponseHeaders()), payload, requestData);
          }

          pendingRequestCount = pendingRequestCount - 1;

          run.join(null, reject, { payload: payload, textStatus: textStatus, jqXHR: jqXHR, errorThrown: errorThrown, response: response });
        };

        pendingRequestCount = pendingRequestCount + 1;

        (0, _emberAjaxUtilsAjax['default'])(hash);
      }, 'ember-ajax: ' + hash.type + ' ' + hash.url);
    },
```

**Comment by fiftoine** [comment]
I'm having the same issue. Is there any news on this? Is there an alternative? 
I use ember-ajax to automatically add a JWT header.

**Comment by alexlafroscia** [comment]
Hey, if you can provide a failing test case or app demo app that reproduces the bug, I'm happy to fix it.

**Comment by XaserAcheron** [comment]
I think this is the same thing as #158. The empty response is being coerced into `{}` by `ember-ajax` (a confirmed bug), and `{}` isn't a valid JSON API document. `ember-data`'s `JSONAPIAdapter.normalizeResponse` doesn't like this, hence the error.

I haven't written an explicit test for this yet since it's essentially a duplicate, but eventually I'll be putting together a test suite for the `AjaxSupport` mixin to make sure that `ember-data` requests and responses remain the same before and after the mixin is applied. This seems like a great test case for that.

**Comment by alexlafroscia** [comment]
Is this closed, now that #232 has been merged?

**Comment by XaserAcheron** [comment]
I believe so, though I never did write an explicit test case for this that uses `ember-data`. May be worth pinging @jerel and @fiftoine just in case.

**Comment by sandstrom** [comment]
@alexlafroscia If I read #232 correctly it doesn't change much (`isNone` is the same as checking for undefined and null). So I don't think it would have affected this.

**Comment by XaserAcheron** [comment]
There's two changed lines in #232, and the real important one is sneakily the *removal* of line 538 [here](https://github.com/ember-cli/ember-ajax/pull/232/files#diff-7002051509692f431bb29f015fcaf32aL538) -- that was causing `ember-ajax` to transform null responses to `{}`, and the corresponding isNone check on line 850 is just there to stop `normalizeErrorResponse` from throwing an error if the newly-allowed null result occurs.

tl;dr: this ought to be fixed, though @jerel and @fiftoine have the most say on that.

**Comment by machty** [comment]
This is still an issue on 3.0.0 (I'm using AjaxServiceSupport)

```
jQuery.parseJSON (vendor-e4e8e9797d8fe3caf3447bc7db0ea473.js:12983)
ajaxConvert (vendor-e4e8e9797d8fe3caf3447bc7db0ea473.js:13260)
done (vendor-e4e8e9797d8fe3caf3447bc7db0ea473.js:13676)
(anonymous) (vendor-e4e8e9797d8fe3caf3447bc7db0ea473.js:14066)
```

**Comment by runspired** [comment]
ember-ajax should be made to return `{ data: null }` in this scenario.
